modules/i18n/input-method/default.nix

```
@@ -0,0 +1,68 @@
+{ config, pkgs, lib, ... }:
+
+with lib;
+let
+  cfg = config.i18n.inputMethod;
+
+  gtk2_cache = pkgs.runCommand "gtk2-immodule.cache" {
+    preferLocalBuild = true;
+    allowSubstitutes = false;
+    buildInputs = [ pkgs.gtk2 cfg.package ];
+  } ''
+    mkdir -p $out/etc/gtk-2.0/
+    GTK_PATH=${cfg.package}/lib/gtk-2.0/ gtk-query-immodules-2.0 > $out/etc/gtk-2.0/immodules.cache
+  '';
+
+  gtk3_cache = pkgs.runCommand "gtk3-immodule.cache" {
+    preferLocalBuild = true;
+    allowSubstitutes = false;
+    buildInputs = [ pkgs.gtk3 cfg.package ];
+  } ''
+    mkdir -p $out/etc/gtk-3.0/
+    GTK_PATH=${cfg.package}/lib/gtk-3.0/ gtk-query-immodules-3.0 > $out/etc/gtk-3.0/immodules.cache
+  '';
+
+in {
+  options.i18n = {
+    inputMethod = {
+      enabled = mkOption {
+        type = types.nullOr
+          (types.enum [ "fcitx" "fcitx5" "nabi" "uim" "hime" "kime" ]);
+        default = null;
+        example = "fcitx";
+        description = ''
+          Select the enabled input method. Input methods is a software to input symbols that are not available on standard input devices.
+
+          Input methods are specially used to input Chinese, Japanese and Korean characters.
+
+          Currently the following input methods are available in NixOS:
+
+          <itemizedlist>
+          <listitem><para>fcitx: A customizable lightweight input method, extra input engines can be added using <literal>i18n.inputMethod.fcitx.engines</literal>.</para></listitem>
+          <listitem><para>fcitx5: The next generation of fcitx, addons (including engines, dictionaries, skins) can be added using <literal>i18n.inputMethod.fcitx5.addons</literal>.</para></listitem>
+          <listitem><para>nabi: A Korean input method based on XIM. Nabi doesn't support Qt 5.</para></listitem>
+          <listitem><para>uim: The universal input method, is a library with a XIM bridge. uim mainly support Chinese, Japanese and Korean.</para></listitem>
+          <listitem><para>hime: An extremely easy-to-use input method framework.</para></listitem>
+          <listitem><para>kime: Koream IME.</para></listitem>
+          </itemizedlist>
+        '';
+      };
+
+      package = mkOption {
+        internal = true;
+        type = types.nullOr types.path;
+        default = null;
+        description = ''
+          The input method method package.
+        '';
+      };
```

```suggestion
      package = mkOption {
        internal = true;
        readOnly = true;
        type = types.nullOr types.path;
        default = null;
        description = ''
          The input method method package.
        '';
      };
```

The `readOnly` attribute is optional but makes it such that the option can only be set once and cannot be overridden.

---

That makes my test fail with the error ``The option `i18n.inputMethod.package' is read-only, but it's set multiple times. Definition values: default.nix: null, kime.nix, uim.nix, nabi.nix, hime.nix, fcitx.nix and fcitx5.nix.

---

Nevermind then, I assumed marking an option `readonly` would ignore `mkIf`/`mkMerge` but it doesn't ¯\\\_\(ツ\)\_/¯